### PR TITLE
AI Assistant: Extend list block and enhance content parsing

### LINKED ISSUE #31125: AI Assistant: sometimes the block generates lists with broken markup
When the assistant is prompted to generate a list of items, sometimes the result in the editor looks like this:

Screen Shot 2023-05-31 at 18 35 06

Exploring the markup for the inserted content, I found `<p>` elements inside `<li>` elements, what can explain the broken visualization.

Screen Shot 2023-05-31 at 18 35 44

### LINKED ISSUE #31875: AI Extensions: extend List Block
We stopped extending List Items now that we convert the extended blocks to the AI Assistant block whenever a request is sent. This is because an AI Assistant can't be nested inside a List block. 

Let's work in extending the List block so lists can be modified using our extension.

### PR DESCRIPTION
**Pull request by dhasilva**
Fixes #31875
Fixes #31125
Part of #31878

## Proposed changes:

* Adds `core/list` to the extended blocks
* Breaks down the `useTextContentFromSelectedBlocks` hook in two for better separation of concerns
* Fixes the broken lists by processing the generated markdown
* Removes the dependency on a blocks `content` attribute, reading its HTML content directly instead

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Testing instructions:

* Add a list and some paragraphs
* Check that you can use the `AI Assistant` toolbar menu on lists
* Check that, when transforming multiple paragraphs, their content is not concatenated into one paragraph, remaining separate

https://github.com/Automattic/jetpack/assets/8486249/e24b6618-5faa-4d72-8be1-32e231056a4c

### CODE REVIEW

File: projects/plugins/jetpack/extensions/blocks/ai-assistant/hooks/use-selected-blocks/index.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+	const blocks = useSelect(
+		select =>
+			( select( 'core/block-editor' ) as BlockEditorStore[ 'selectors' ] ).getBlocksByClientId(
+				clientIds
+			),
+		[ clientIds ]
+	);
+
+	const defaltValue = {
```
Comment (nunyvega) [inline projects/plugins/jetpack/extensions/blocks/ai-assistant/hooks/use-selected-blocks/index.ts]:
```suggestion
	const defaultValue = {
```

File: projects/plugins/jetpack/extensions/blocks/ai-assistant/hooks/use-selected-blocks/index.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+		count: 0,
+		clientIds: [],
+		blocks: [],
+	};
+
+	if ( ! clientIds?.length || ! blocks?.length ) {
+		return defaltValue;
```
Comment (nunyvega) [inline projects/plugins/jetpack/extensions/blocks/ai-assistant/hooks/use-selected-blocks/index.ts]:
```suggestion
		return defaultValue;
```

**Review by nunyvega — APPROVED** [review]
Works well!

🚢 
:shipit:

File: projects/plugins/jetpack/extensions/blocks/ai-assistant/extensions/ai-assistant/with-ai-assistant.tsx [hunk 3]
```diff
@@ -189,19 +189,22 @@ export const withAIAssistant = createHigherOrderComponent(
...
 			);
 
 			removeBlocks( otherBlocksIds );
 		}, [ blocks, blockType, content, replaceBlock, clientIds, removeBlocks ] );
 
-		const rawContent = getRawTextFromHTML( props.attributes.content );
+		const blockControlProps = {
+			group: 'block',
+		};
+		const rawContent = getRawTextFromHTML( content );
```
Comment (ellatrix) [inline projects/plugins/jetpack/extensions/blocks/ai-assistant/extensions/ai-assistant/with-ai-assistant.tsx:201]: Are you running regexes, and setting innerHTML on every render of this block higher order component? https://github.com/Automattic/jetpack/pull/31901/files#diff-6dbc088debefb5e5c987ea76a69651132bf1c7d2234fd16ac3cbf0815c88ed20R95

Comment (ellatrix) [reply projects/plugins/jetpack/extensions/blocks/ai-assistant/extensions/ai-assistant/with-ai-assistant.tsx:201]: This higher order component is extending every block probably? So this is running on every render for every block in a post?

Comment (ellatrix) [reply projects/plugins/jetpack/extensions/blocks/ai-assistant/extensions/ai-assistant/with-ai-assistant.tsx:201]: Not every block, but pretty much every block https://github.com/Automattic/jetpack/blob/5daa8f1be69422fd0106a1b6a7fbb5c62105ed8c/projects/plugins/jetpack/extensions/blocks/ai-assistant/extensions/ai-assistant/index.ts#L20

Comment (dhasilva) [reply projects/plugins/jetpack/extensions/blocks/ai-assistant/extensions/ai-assistant/with-ai-assistant.tsx:201]: That is correct. Issues being addressed on https://github.com/Automattic/jetpack/pull/33681
ref p1697714798422019-slack-C02TQF5VAJD

**Pull request merged**